### ISSUE TITLE
加上reset_net后loss丢失

### ISSUE THREAD

**Issue by WeihongPan** [fangwei123456/spikingjelly#232]
用的是最新的SJ，在vgg16的5层卷积后接上如下全连接层
```python
nn.Flatten(),
nn.Linear(512*30*40, 256, bias=False),
neuron.IFNode(surrogate_function=surrogate.ATan()),
nn.Linear(256, 64, bias=False),
neuron.IFNode(surrogate_function=surrogate.ATan()),
```

损失函数为triplet loss（nn.TripletMarginLoss），在两张gpu上训练，但一轮iteration后reset_net(model)，导致下一轮loss丢失：
```bash
... 
===>Query  3 , Neg  8
d_pos:  tensor([0.5892], device='cuda:0', grad_fn=<SumBackward1>)
d_neg:  tensor([0.4804], device='cuda:0', grad_fn=<SumBackward1>)
loss_i:  tensor(0.4250, device='cuda:0', grad_fn=<MeanBackward0>)
triplet loss:  tensor(0.3907, device='cuda:0', grad_fn=<SumBackward0>)
===>Query  3 , Neg  9
d_pos:  tensor([0.5892], device='cuda:0', grad_fn=<SumBackward1>)
d_neg:  tensor([0.4835], device='cuda:0', grad_fn=<SumBackward1>)
loss_i:  tensor(0.4219, device='cuda:0', grad_fn=<MeanBackward0>)
triplet loss:  tensor(0.3885, device='cuda:0', grad_fn=<SumBackward0>)
loss:  tensor(14.3343, device='cuda:0', grad_fn=<AddBackward0>)
loss after normalization:  tensor(0.3584, device='cuda:0', grad_fn=<DivBackward0>)
=====> iteration 2
negCounts:  tensor([10, 10, 10, 10])
===>Query  0 , Neg  0
d_pos:  tensor([0.], device='cuda:0', grad_fn=<SumBackward1>)
d_neg:  tensor([0.], device='cuda:0', grad_fn=<SumBackward1>)
loss_i:  tensor(0.3162, device='cuda:0', grad_fn=<MeanBackward0>)
triplet loss:  tensor(0.3162, device='cuda:0', grad_fn=<SumBackward0>)
===>Query  0 , Neg  1
d_pos:  tensor([0.], device='cuda:0', grad_fn=<SumBackward1>)
d_neg:  tensor([0.], device='cuda:0', grad_fn=<SumBackward1>)
loss_i:  tensor(0.3162, device='cuda:0', grad_fn=<MeanBackward0>)
triplet loss:  tensor(0.3162, device='cuda:0', grad_fn=<SumBackward0>)
===>Query  0 , Neg  2
d_pos:  tensor([0.], device='cuda:0', grad_fn=<SumBackward1>)
d_neg:  tensor([0.], device='cuda:0', grad_fn=<SumBackward1>)
loss_i:  tensor(0.3162, device='cuda:0', grad_fn=<MeanBackward0>)
triplet loss:  tensor(0.3162, device='cuda:0', grad_fn=<SumBackward0>)
...
=====> iteration 3
negCounts:  tensor([10, 10, 10, 10])
===>Query  0 , Neg  0
d_pos:  tensor([nan], device='cuda:0', grad_fn=<SumBackward1>)
d_neg:  tensor([nan], device='cuda:0', grad_fn=<SumBackward1>)
loss_i:  tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
triplet loss:  tensor(nan, device='cuda:0', grad_fn=<SumBackward0>)
===>Query  0 , Neg  1
d_pos:  tensor([nan], device='cuda:0', grad_fn=<SumBackward1>)
d_neg:  tensor([nan], device='cuda:0', grad_fn=<SumBackward1>)
loss_i:  tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
triplet loss:  tensor(nan, device='cuda:0', grad_fn=<SumBackward0>)
===>Query  0 , Neg  2
d_pos:  tensor([nan], device='cuda:0', grad_fn=<SumBackward1>)
d_neg:  tensor([nan], device='cuda:0', grad_fn=<SumBackward1>)
loss_i:  tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
triplet loss:  tensor(nan, device='cuda:0', grad_fn=<SumBackward0>)
```

训练函数如下：
```python
input = input.to(device)
image_encoding = model.encoder(input) # (24, 512, 30, 40)
vlad_encoding = model.pool(image_encoding) # (24, 64, 512)

vladQ, vladP, vladN = torch.split(vlad_encoding, [B, B, nNeg]) 
optimizer.zero_grad()
            
# calculate loss for each Query, Positive, Negative triplet
# due to potential difference in number of negatives have to 
# do it per query, per negative
loss = 0
for i, negCount in enumerate(negCounts): # negCounts=[10,10,10,10]
    for n in range(negCount):
    negIx = (torch.sum(negCounts[:i]) + n).item()
    # debug
    d_pos = torch.sum(torch.square(vladQ[i:i+1] - vladP[i:i+1]), 1)
    d_neg = torch.sum(torch.square(vladQ[i:i+1] - vladN[negIx:negIx+1]), 1)
    loss_i = torch.maximum(torch.tensor(0.0), opt.margin**0.5 + d_pos - d_neg)
    loss_i = torch.mean(loss_i)
    print('===>Query ',i,', Neg ', n)
    print('d_pos: ', d_pos)
    print('d_neg: ', d_neg)
    print('loss_i: ', loss_i)
    print('triplet loss: ', criterion(vladQ[i:i+1], vladP[i:i+1], vladN[negIx:negIx+1]))

    loss += criterion(vladQ[i:i+1], vladP[i:i+1], vladN[negIx:negIx+1])
    # criterion = nn.TripletMarginLoss(margin=opt.margin**0.5, p=2, reduction='sum').to(device)
print('loss: ', loss)
loss /= nNeg.float().to(device) # normalise by actual number of negatives
print('loss after normalization: ', loss)

loss.backward()
optimizer.step()        
batch_loss = loss.item()
epoch_loss += batch_loss
functional.reset_net(model)
```

**Comment by fangwei123456** [comment]
输出一下iteration 3的vlad_encoding看看

**Comment by WeihongPan** [comment]
```bash
vlad encoding:  tensor([[nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        ...,
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan]], device='cuda:0',
       grad_fn=<GatherBackward>)
```
输出发现是加上IFNode之后vgg出现梯度爆炸的问题（e+9~e+10），导致iteration 1开始encoder输出值巨大e+27~e+28数量级

**Comment by fangwei123456** [comment]
看起来并非reset导致，应该是训练本身的问题
